loader/src/lib.rs

```diff
@@ -1,43 +1,59 @@
 // SPDX-License-Identifier: Apache-2.0
 
+//! A helper crate for loading ELF binaries. These can
+//! be referred to as "components" in Enarx parlance
+//! (for example: shim and code components).
+
+#![deny(clippy::all)]
+#![deny(missing_docs)]
+
 use goblin::elf::{header::*, program_header::*, Elf};
 
 use memory::Page;
-use sgx_types::page::{Flags, SecInfo};
 use span::{Line, Span};
 
 use std::cmp::{max, min};
 use std::io::Result;
 use std::ops::Range;
 use std::path::Path;
 
+/// The Segment parsing logic provided by this crate
+/// allows client code to optionally participate in
+/// reading the ELF program headers during parsing.
+/// Client code can have a struct `impl FromHeader`
+/// to optionally store additional information in
+/// the Segment<T> struct.
+///
+/// This can be opted out of by using `Component<()>`.
+pub trait FromHeader {
+    /// A constructor function to help client code
+    /// store additional information from the ELF
+    /// segment.
+    fn from_header(ph: &ProgramHeader) -> Self;
+}
+
+impl FromHeader for () {
+    fn from_header(_ph: &ProgramHeader) -> Self {}
+}
+
 /// A loadable segment of code
-pub struct Segment {
+pub struct Segment<T: FromHeader> {
```

Comment:
Rather than this approach, I think we should make two separate segment types. One for the `Builder` and one for the `loader`. Then `enarx-keep-sgx-shim` can carry a function to convert between the two. In particular, I want to avoid a dependency between the `loader` and the `Builder`. This way when we merge the `sgx` libraries, I can move the `Builder` code from here to there without adding a dependency.